### PR TITLE
Rule S2761: "Not(...)" boolean operator should not be doubled

### PR DESCRIPTION
**Pull request by valhristov**

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UnaryPrefixOperatorRepeatedBase.cs [hunk 1]
```diff
@@ -1,91 +1,94 @@
...
+                        currentUnary = GetOperand(currentUnary) as TSyntaxNode;
+                    }
+
+                    if (repeatedCount < 2)
+                    {
+                        return;
+                    }
+
+                    var errorLocation = new TextSpan(
+                        topLevelUnary.SpanStart,
```
[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UnaryPrefixOperatorRepeatedBase.cs]: is this FP?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/VisualBasicUnaryPrefixOperatorRepeated.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class VisualBasicUnaryPrefixOperatorRepeated :
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/VisualBasicUnaryPrefixOperatorRepeated.cs]: why do we need the language prefix (`VisualBasic`) - it's clear from the namespace what it is

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.Fixed.cs [hunk 3]
```diff
@@ -6,12 +6,13 @@ namespace Tests.Diagnostics
 
     class UnaryPrefixOperatorRepeated
     {
-        static void NonComp(bool b  )
+        static void NonComp(bool bbb)
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.Fixed.cs]: is this `bbb` intended?

Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.Fixed.cs]: is `bbb` used?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.cs [hunk 4]
```diff
@@ -6,13 +6,14 @@ namespace Tests.Diagnostics
 
     class UnaryPrefixOperatorRepeated
     {
-        static void NonComp(bool b  )
```
[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.cs]: you could remove `bbb` altogether

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.vb [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿Imports System
+Imports System.Collections.Generic
+
+Namespace Tests.Diagnostics
+    Class UnaryPrefixOperatorRepeated
+        Private Shared Sub NonComp(ByVal bbb As Boolean)
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.vb]: same here, `bbb` can be removed

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/VisualBasicUnaryPrefixOperatorRepeated.cs (hunk 2 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/VisualBasicUnaryPrefixOperatorRepeated.cs]: For some reason my Visual Studio has hard time navigating through classes that have the same name (it has been like this for long time and I reinstalled it completely yesterday)... For example, if a compiler error appears in one of the files and you double click it in the output window, VS navigates to the other file.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.Fixed.cs (hunk 3 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.Fixed.cs]: `bbb` is used a few lines below (the same method is called recursively)

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.vb (hunk 5 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UnaryPrefixOperatorRepeated.vb]: `bbb` is used a few lines below (the same method is called recursively)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UnaryPrefixOperatorRepeatedBase.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        where TSyntaxKindEnum : struct
+    {
+        internal const string DiagnosticId = "S2761";
+        protected const string MessageFormat = "Use the '{0}' operator just once or not at all.";
+
+        protected abstract DiagnosticDescriptor Rule { get; }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected abstract HashSet<TSyntaxKindEnum> SyntaxKinds { get; }
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UnaryPrefixOperatorRepeatedBase.cs]: this can be an `ISet`

**Pull request merged**